### PR TITLE
Issue #2379: portfolio domain readonly view

### LINKED ISSUE #2379: Domain Management pages Permission handling and read-only view
### Issue description

On the domain management pages, we will have users who only have the rights to **see** domain data and to not edit it. This is to give them insight into their domains without risking them changing things they aren't supposed to. We already have an editable portfolio view, and now we need that view to only show if the user has management permissions and change into a non edit view if the user has edit permissions. 

### Acceptance criteria

- [ ] The[ edit view ](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-Designs-CISA?node-id=4788-13705&t=XYW1P7m3CAtY7vWq-0) created in #2354 is only shown if a user has the permissions to manage a domain
- [ ] the [read-only view](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-Designs-CISA?node-id=4089-43032&t=XYW1P7m3CAtY7vWq-0) is shown if a user **only** has view permissions to the domains (such as admins and read-only admins)
    - [ ] Note, the info message change and side nav change in this view

### Additional context

permissions were handled in #2350 for the Domain table, follow what was done there. 

### Links to other issues

blocked by #2350

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #2379 

## Changes

- Readonly view for domain page added when user has PortfolioDomain permission but is not a Domain Manager for a domain.  User can View domain.

## Code Review Verification Steps

Make sure waffle flag for organization_feature is set to Yes
Add a portfolio
Set your user's portfolio to the portfolio you just created, and grant yourself Admin portfolio role
For two or more domains, set the portfolio of the domain to the portfolio you created
For at least one of the domains, add a UserDomainRole for yourself.  For at least one of the domains, don't add a UserDomainRole for yourself.
When you navigate next to the home page, you should see a mixture of domains with View and Manage, based on whether you are a DomainManager for the domain.
Click View.  You should see the requested readonly view of a domain.

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/templates/domain_detail.html
Comment (dave-kennedy-ecs) [inline src/registrar/templates/domain_detail.html:43]: display alert for portfolio users who are not domain managers

File: src/registrar/templates/domain_detail.html
Comment (dave-kennedy-ecs) [inline src/registrar/templates/domain_detail.html:58]: This and many other checks in the template needed to be changed from editable based solely on the state of the domain to be based on the state of the domain plus aspects of the user, so wrote a new function for that at the view level.

File: src/registrar/context_processors.py
Comment (dave-kennedy-ecs) [inline src/registrar/context_processors.py:69]: portfolio context removed from individual views get_context_data since context effects many views across the app and therefore is added globally in context_processors

File: src/registrar/views/portfolios.py
Comment (abroddrick) [inline src/registrar/views/portfolios.py:24]: so much cleaner honestly

File: src/registrar/views/portfolios.py
Comment (abroddrick) [inline src/registrar/views/portfolios.py:54]: **minor suggestion/non blocking** a function on user for get_portfolio that returns the id may make this more readable (plus I imagine getting the portfolio ID is something we would need to do often. Not a must though. Your code is so clean, I have to comment on something

File: src/registrar/views/utility/mixins.py
Comment (abroddrick) [inline src/registrar/views/utility/mixins.py:194]:
```suggestion
        If particular views allows access to the domain pages, they will need to override
```

**suggestion not blocking**

**Review by abroddrick — COMMENTED** [review]
Still have a few more things to check on here, but submitting comments for now. Just that one blocking comment

**Comment by erinysong** [comment]
Will defer to design on this but was curious if the search and filter elements should be more spaced apart from each other

image
image

**Comment by gabydisarli** [comment]
> Will defer to design on this but was curious if the search and filter elements should be more spaced apart from each other
> 
> image image

Yes, we have it as 24px in the design

**Comment by abroddrick** [comment]
Thanks @erinysong for sharing my blocking comment! I apparently was trying to go through the review to fast

**Comment by abroddrick** [comment]
@dave-kennedy-ecs all looks good minus the view icon not showing up and the spacing that erin found, all other comments are mino/optional so far. @erinysong please also keep looking at this PR if you're still reviewing, it's one that is AMAZING to have two eyes on!

**Review by erinysong — COMMENTED** [review]
Also am good to approve once the icon and spacing comments from above are updated! Verified that read and write views work and code looks good

**Comment by dave-kennedy-ecs** [comment]
> Will defer to design on this but was curious if the search and filter elements should be more spaced apart from each other
> 
> image image

I am pretty sure that @zandercymatics was fixing this in a different PR

File: src/registrar/views/portfolios.py
Comment (dave-kennedy-ecs) [reply src/registrar/views/portfolios.py:54]: > **minor suggestion/non blocking** a function on user for get_portfolio that returns the id may make this more readable (plus I imagine getting the portfolio ID is something we would need to do often. Not a must though. Your code is so clean, I have to comment on something

So far, this is the only place where we are retrieving the portfolio id.  There are three other places in code where we are getting portfolio from user.  I am not sure yet if adding another method for getting the portfolio from the user is simpler than just user.portfolio.  Is the idea that portfolio might be None?

**Comment by dave-kennedy-ecs** [comment]
> **blocking** Just logged in to the system and see this icon on VIew only ones. This should be the eye icon instead when it is view only

OK, fixed this!  Good catch, thanks

**Comment by dave-kennedy-ecs** [comment]
> Will defer to design on this but was curious if the search and filter elements should be more spaced apart from each other

Yes, this was fixed by a different PR from @zandercymatics , so I have merged main in to take advantage of the fixed style.  It should be working as expected now

File: src/registrar/views/portfolios.py
Comment (dave-kennedy-ecs) [reply src/registrar/views/portfolios.py]: Ok, decided to fix this so that it doesn't depend on the id of the portfolio.  It just returns the portfolio, which is simpler.  And still raises 404 if portfolio is none

File: src/registrar/context_processors.py
Comment (dave-kennedy-ecs) [reply src/registrar/context_processors.py:78]: I don't know.  I think this is a good thought.  I was honestly hesitant to use global context for anything that hits the database because the load it would put on every request.  But decided to do this since these variables need to be available in context for most user facing views.

**Review by erinysong — APPROVED** [review]
LGTM!

**Review by abroddrick — APPROVED** [review]
LGTM

**Pull request merged**
